### ISSUE TITLE
[Bug] AutoEP FP32 gradient clipping averages norms across ranks with different expert ownership

### ISSUE THREAD

**Issue by gss10282023** [deepspeedai/DeepSpeed#8469]
### Summary

DeepSpeed's FP32 gradient-clipping path computes the wrong global norm when AutoEP places different experts on different ranks. It averages rank-local norms as if every rank held the same parameters, but under AutoEP the ranks own different expert parameters.

Redistributing the same unique expert gradients across ranks can therefore change the clipping coefficient and the resulting update when clipping is active.

### System info

**Environment:** DeepSpeed commit [cf443004](https://github.com/deepspeedai/DeepSpeed/blob/cf44300453eb0af79ed84ed8f1cb49d57478bd76/deepspeed/runtime/utils.py), two CUDA ranks, FP32 gradients, ZeRO-0, L2 clipping, `mpu=None`.

The results below are from the pinned version. Source inspection on 2026-09-10 found the same calculation in [0166cc87](https://github.com/deepspeedai/DeepSpeed/blob/0166cc878b106546adc78214da952f1aa6f5d02b/deepspeed/runtime/utils.py); the full experiment has not been rerun on that commit.

Separate current-branch CPU validation: Ubuntu 26.04 LTS (x86_64), Python 3.11.15, PyTorch 2.13.0+cpu, Intel Xeon Cascadelake, two local CPU/Gloo processes, source baseline `0166cc87` and PR #8470. No GPU or inter-node connection is used in that CPU run.

### To Reproduce

The test enters through `DeepSpeedEngine.clip_fp32_gradients()` and represents the same four unique expert gradients in two layouts:

- **EP size 1:** all four experts replicated on both ranks.
- **EP size 2:** each rank owns two of the four experts.

The unique gradient vector and the clipping threshold are fixed. The expected coefficient comes directly from the L2 norm of the unique vector, counting each expert once.

| Comparison | Absolute difference in clipping coefficient |
| --- | ---: |
| Replicated vs. owner-sharded, current code | `0.2000005` |
| Replicated vs. owner-sharded, with repair | `0` |
| Fully replicated control | `0` |

The replicated layout matches the direct reference; the owner-sharded layout does not. The discrepancy is identical across three repeated blocks, and the repair was confirmed on a second host.

The regression code is in [PR #8470](https://github.com/deepspeedai/DeepSpeed/blob/8288645c1cef4c03bb65fb316fb6473bcb13f02f/tests/unit/v1/moe/test_autoep_grad_parity.py). `TestAutoEPFP32Clipping` exercises the native engine against an independent unique-gradient reference. [PR #8470](https://github.com/deepspeedai/DeepSpeed/pull/8470) includes the exact two-process CPU/Gloo command; all eight cases passed with the repair, and the owner-sharded case fails on unmodified `0166cc87`.

### Expected behavior

Replicated and owner-sharded layouts of the same unique gradients should give the same global L2 norm, clipping coefficient, and optimizer update. Dense parameters should count once despite replication, and each expert should count once according to its ownership.

### Root cause

`clip_grad_norm_` takes a local square root before averaging across the data-parallel group. When ranks own different experts, the average of local norms does not reconstruct the norm of the concatenated unique gradients.

Replication counts differ as well: dense parameters live on every DP rank, while AutoEP expert parameters live on only `dp_world_size / ep_size` ranks. Their contributions need different weights for each logical parameter to count exactly once.

### Proposed fix

Record the EP size on each AutoEP expert parameter. For the FP32 L2 path:

1. Sum squared gradients locally, dividing each parameter's contribution by its replica count.
2. Sum those contributions across the DP group.
3. Take a single square root after the reduction and compute the clipping coefficient.

The tested patch adds the ownership metadata in `auto_ep_layer.py` and implements this reduction in `runtime/utils.py`. Its validated scope is the ZeRO-0, `mpu=None`, L2 case above.

### ds_report output

Collected with `DS_ACCELERATOR=cpu python -m deepspeed.env_report` from the current PR checkout. This describes the separate CPU validation environment; it is not the pinned two-GPU environment. Local installation paths have been replaced with placeholders.

<details>
<summary>CPU environment report</summary>

```text
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
deepspeed_not_implemented  [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
deepspeed_ccl_comm ..... [NO] ....... [OKAY]
deepspeed_shm_comm ..... [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
pin_memory ............. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['<local dependency path>']
torch version .................... 2.13.0+cpu
deepspeed install path ........... ['<DeepSpeed checkout>/deepspeed']
deepspeed info ................... 0.19.6, [none], [none]
deepspeed wheel compiled w. ...... torch 0.0 
shared memory (/dev/shm) size .... 30.27 GB
```

</details>

### Launcher context

The current regression uses `python -m torch.distributed.run --standalone --nproc-per-node=2` with Gloo. The exact driver is included in PR #8470.

### Docker context

The current CPU regression ran in an existing Ubuntu environment; no separately specified Docker image is required by the reproducer. The pinned GPU report does not specify an image tag.

**Comment by alanhuangyoo** [comment]
@gss10282023 Independently reproduced, and #8478 is up.

I came at it from a different direction than your harness — four fixed expert gradients, two ranks, reading the norm directly rather than the clipping coefficient — and it lands in the same place:

```
reference, each expert counted once   10.954452     sqrt(120)
replicated on both ranks              10.954452
one half per rank                      7.236068     (sqrt(20) + sqrt(100)) / 2
```

So your 0.2 coefficient shift and this 34% norm gap are the same defect seen through two different lenses.

One thing your report did not mention that I think strengthens it: **the repo already contains both implementations.** `get_norm_with_moe_layers` is the correct combiner and is used by the bf16 and fp16 optimizers. The approximation also exists under its own name, `get_norm_with_moe_layers_fast`, carrying this comment:

> This implementation standardizes the grad_norm across ranks. A more precise implementation can be found in 'get_norm_with_moe_layers'.

and that named copy has **no callers**. `clip_grad_norm_` inlines the same logic without the comment, so the one place still doing the approximate thing is the one place that never got the warning.

That is the route #8478 takes: keep expert gradients out of the replicated norm, then fold them in through the precise helper, so fp32 matches the other two optimizers instead of carrying a second implementation.

I noticed you closed #8470 without comment. If you withdrew it because you found a problem with the approach, I would rather hear it now than have it rediscovered in review — and if you would prefer to carry this yourself, say so and I will close mine.

**Comment by gss10282023** [comment]
@alanhuangyoo Thanks for the independent reproduction. I closed #8470 during resubmission; the current report and fix are #8475 and #8476.

#8478 reuses the existing helper and covers legacy MoE; #8476 targets AutoEP FP32 L2. Both passed the common CPU/Gloo clipping checks, including mesh-SP. At `875e8fe`, #8478 also showed a collective hang when one rank had no expert gradients, and an exception for expert-only infinity-norm clipping.

I can share these cases so we can work together on a unified fix.

**Comment by alanhuangyoo** [comment]
@gss10282023 Both of those are real, and I reproduced them on my own branch before replying rather than taking them on trust. Thank you — the hang in particular is the kind of thing that would have been found by someone's training run rather than by review.

**Uneven expert gradients — deadlock.** rank 0 holding an expert gradient, rank 1 holding none, 2×H20:

```
[rank0] entering clip_grad_norm_ (has expert grads)
[rank1] entering clip_grad_norm_ (no expert grads)
<neither returns; killed at 90s>
```

The cause is mine and it is a design error, not an oversight in a corner: I branch on whether **this rank** has expert gradients, and the two branches issue different collectives. A rank whose experts got no tokens this step takes the averaging path while its peer takes the expert path, and they wait on each other. `p.grad is not None` filtering makes that an ordinary MoE situation, not a rare one.

**Expert-only inf norm — exception.** With every parameter an expert and `norm_type=inf`, `norm_parameters` is empty and `torch.stack([])` raises.

The fix for the first has to make the branch decision rank-independent. `_EXPERT_PARALLEL_GROUP` is built identically on every rank, so iterating the registry's group names — and contributing a zero tensor for a name this rank holds nothing for — gives every rank the same collectives regardless of which experts got tokens. The second is a guard on the empty list, with 0 as the identity for both `max` and the p-sum.

**On working together: yes.** Reading #8475/#8476 now. As I understand the split, #8476 repairs the AutoEP FP32 L2 computation, and #8478 routes the fp32 path through `get_norm_with_moe_layers`, which is what the bf16 and fp16 optimizers already call. Those are not the same change and one of them may subsume the other, so rather than land two overlapping fixes:

- if your reproductions show the cases you found also apply to the helper-based route, I would rather fold my route into yours than carry a second implementation;
- if the helper route is the one to keep, the cases you found belong in it as tests, and I would want them from you rather than reconstructing them.

Please do share the cases. If a maintainer prefers a single PR, I am happy for it to be yours — you reported this and had a fix up first.
